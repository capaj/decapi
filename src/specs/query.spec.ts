import { graphql, GraphQLSchema, printSchema } from 'graphql'
import {
  compileSchema,
  ObjectType,
  Field,
  SchemaRoot,
  Query
} from '../index.js'

@ObjectType()
class Hello {
  @Field()
  world(name: string): string {
    return `Hello, ${name}`
  }

  @Field()
  categories(): Promise<string[]> {
    return Promise.resolve(['Tables', 'Furniture'])
  }
}

@SchemaRoot()
class FooSchema {
  @Query()
  hello(): Hello {
    return new Hello()
  }

  @Query()
  foo(): string {
    return 'bar'
  }
}

describe('Query', () => {
  let schema: GraphQLSchema
  it('should compile', () => {
    schema = compileSchema(FooSchema)
    expect(printSchema(schema)).toMatchSnapshot()
  })

  it('should support queries with simple arguments', async () => {
    const result = await graphql({
      schema,
      source: `
        {
          hello {
            world(name: "Bob")
            categories
          }
        }
      `
    })

    expect(result).toMatchSnapshot()
  })

  it('should not allow wrong argument types', async () => {
    const result = await graphql({
      schema,
      source: `
        {
          hello {
            world(name: 2)
          }
        }
      `
    })
    expect(result.errors).toBeDefined()
    expect(result.errors).toMatchSnapshot()
  })

  it('will support flat root fields', async () => {
    const result = await graphql({
      schema,
      source: `
        {
          foo
        }
      `
    })
    expect(result).toEqual({ data: { foo: 'bar' } })
  })
})
