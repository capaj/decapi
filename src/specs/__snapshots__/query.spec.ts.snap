// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query should not allow wrong argument types 1`] = `
Array [
  [GraphQLError: String cannot represent a non string value: 2],
]
`;
